### ISSUE TITLE
GGMAX Enhancement - Real Empty Levels

### ISSUE THREAD

**Issue by OldFlak** [TheGameCreators/GameGuruRepo#5827]
Please add ability to have proper empty levels.
Currently an empty level is just flat terrain - so really it should be named just that - **Flat terrain**

I have a space level created using the so-called Empty Level option - with no assets added and just using a Start Marker and a 2K skybox clocked in at around 4.2.

Now as it is for a space level the terrain is turned off - or rather not rendered - so it is a complete waste of resources.
As it stands the whole level with assets to date is around 7.5 gig of ram usage. More is used by the invisible terrain than the assets.

So can we please have an Empty Level as in **really empty** or at least a single 2 tri plane

Thanks for considering :)

**Comment by mav3r1ck1981** [comment]
Ill second. third and one hundred this! I could finally start my future project, without a truly empty level, its just not practical.

**Comment by C4ever08** [comment]
Yes i agree. It could be added as a tick box to turn off completely. it takes out the immersion in a space game when you hit invisible terrain floor, its forced me to move my assets way up high  this would be beneficial.

**Comment by synchromesh62** [comment]
Its not as easy as it sounds .. 
Right now something needs to be there because the whole Navmesh System relies on it being there, 
If you just turn it off the Navmesh will not function. 
Its a lot of work and not a simple check box unfortunately :(

**Comment by OldFlak** [comment]
Well it is a feature request - not expecting it any time soon, and its more about removing wasted resources i.e. 4+gig of unseen geometry.

That said it may be easier than it seams - just a two try plane that covers the playable area. that gives the navmesh a base, and gets rid of the wasted ram usage for something you can't see.

In GGc you could actually walk off the edge of the map, by adding a bit of code to stop the engine preventing you. 
The endless terrain in MAX is not really endless, just renders stuff outside the playable area. That tho was not really thought through properly as you can get to the edge, and witness Lawnmower Dans activities in culling the grass.

So I would suggest that an empty level also allows you to walk off the edge of the map - other games types would benefit from the illusion of an endless abyss below. After all if you can't go there what is the point of endless terrain and the resources it uses anyway, you can fake that with billboards or skyboxes.

It would not need a checkbox - empty is empty - so that level will not need painted terrain or the ability to sculpt, it will rely completely on your models for walkable areas.

**Comment by firstbastion** [comment]
I like this idea too. There's lots of reasons a real empty level could be useful. And properly resource managed.  And if it can't be empty,  then a small 10x10 terrain just at the player start point could hopefully be a compromise.  It would also have to use just the checker terrain,  because all those 2k terrain textures are probably bloating up the resources.

**Comment by OldFlak** [comment]
Just wondering why this has not been added yet - nothing else is being done from what I can see :)

**Comment by LeeBamberTGC** [comment]
@OldFlak Made a start on this for the next DEVTEST build, a new tick box will be available in the Terrain Generator section that will wipe out the terrain, trees, grass, water, terrain collision, terrain handling. If the project contains no levels that use terrain, then the standalone will also omit the terraintextures, trees and grass folders also (saving about 1GB for a typical one level game).  Check it out when the DEVTEST goes out in a few hours :)

**Comment by OldFlak** [comment]
@LeeBamberTGC - Lee, I can only say Thank You! - never thought this would be realized - huge plus for MAX, and a boost to my plummeting expectations for MAX.

This is fantastic - can't wait to test it out!

**Comment by OldFlak** [comment]
@LeeBamberTGC 
Lee this is a great start - thank you for this!

The other issue we have with space levels is the limited size of a map - 5KM is just not big enough for space levels
- is it possible to up the boundary for empty levels to 50KM

**Comment by C4ever08** [comment]
I second this. I haven't seen the completely empty level yet but It would benefit us making space orientated games to have room for exploration or one level per solar system with the advantage of 50k space. I feel 5km is just enough for one of my planets and a station. It would make a huge difference to the work load of my project and less loading of levels.

**Comment by OldFlak** [comment]
Ideally it would be good to be able to work below 0 in the editor as far as building levels - yes! 
But if that is not doable, or too difficult to add at present then building everything above 0 would work.
- For my part the navmesh is not needed anyway, as I don't intend to use npc's unless they actually become intelligent at some point :)

But a larger area to build in is definitely needed for space levels - 5K is nowhere near big enough, and it is not like every part of a 50km area would be populated with much more than space debris - which can move with the player anyway. So it is really about bigger levels with smaller pockets of models in places.

**Comment by LeeBamberTGC** [comment]
@OldFlak A new DEVTEST build has been uploaded that increases the area from 5K to 50K. I have also added an extra tick box in Terrain Generator to switch off any NAVMESH generation for those levels you do not need them.

The only thing I will probably kick down the road a little is the removal of the ZERO ground level when editing. I find it useful to quickly drag things from the library to the level edit area before placing them around with the widget and having no virtual floor presents some knock-on effects that might mean some fundamental rule changes, and likely to confuse users who are used to have some kind of floor to start from.  I am sure there is a solution for total void editing, but a lot of the editing tools kind of expect a surface to provide some certainty where the object should be placed.

Let me know how the new build goes and if you need anything else before we close this feature request :)

**Comment by OldFlak** [comment]
@LeeBamberTGC  This is great - thanks for that Lee

The zero ground level I can work with - was just a nice to have.
- I guess it is similar to Unity anyhow. 
- A level in Unity has visible grid in the editor, similar 3D tools like Blender, but when you run the level there is nothing there. 
- You add terrain just like any other model, but when you do that they snap to the grid.

But happy for it to left at that - awesome addition!

This feature request is sorted from my perspective, and really appreciated.

When the setskyto Lua issue is sorted (#5902) we can then have players warping to other galaxies all in one map!

**Comment by OldFlak** [comment]
@LeeBamberTGC Hi Lee :)
Not tested this myself, but one of @C4ever08 has a Completely Empty Level in development and is terrain in reflections.
- Is this able to be sorted? :)

**Comment by C4ever08** [comment]
Thank you @OldFlak yes i noticed it as i was putting my station windows in. here is a screenshot of it.

![Image](https://github.com/user-attachments/assets/3f757194-cc67-4006-8599-528d6e4393eb)

**Comment by LeeBamberTGC** [comment]
@C4ever08 I created a super empty level, made a floor a wall and a red ball, and then a puddle splatted on the wall but saw no terrain:

![Image](https://github.com/user-attachments/assets/f5968154-8c94-4333-a128-0dfb92e2cf33)

Can you do the same and confirm you get the same result?  Thanks!

**Comment by C4ever08** [comment]
Yes I apologize. So yes made a brand new 50km map added stock wall and probe. wall is to stand on and it still showing the terrain i also noticed again the bounding box thats usually around the probe is offset like in pic

![Image](https://github.com/user-attachments/assets/6a9dfc78-c797-40a2-aee9-1d3e31743c48)

Down at bottom at walls edge the box is there and small but not where its supposed to be and you can't adjust its settings

![Image](https://github.com/user-attachments/assets/6452fac7-11dd-42bc-813a-8ddc39ec4ea2)

here is the FPM with just walls and probe

[mapbank.zip](https://github.com/user-attachments/files/19822145/mapbank.zip)

**Comment by LeeBamberTGC** [comment]
@C4ever08 Fixed terrain and grass rendering in env probes during completely empty level mode, will be in the next DEVTEST build:

![Image](https://github.com/user-attachments/assets/e02d3f40-a26b-4153-8022-7ac79e9d863d)

**Comment by LeeBamberTGC** [comment]
@C4ever08 An restores env probe refreshes for things like the probe settings and bound box functionality (was lurking inside the terrain updater that was being skipped) :)

**Comment by C4ever08** [comment]
Yes Lee that did the trick probe works and reflections are back to normal Thank you for the quick fix.

**Comment by OldFlak** [comment]
@LeeBamberTGC Looks like this thread is completed.

Only other thing to consider is:
![Image](https://github.com/user-attachments/assets/4c4b96df-5e8d-4756-9103-f5bbdd4b83e4)

But if that is not easily added, and @C4ever08 issue is sorted, then I be happy for this thread to be closed 👍

**Comment by LeeBamberTGC** [comment]
Yay, another feature request completed :)  I will close this and see if the sub-50KM request comes up again for other users. As it's fixed at the start of level creation, it might be easier to leave it MASSIVE to allow future level ideas to expand into it, there is no extra resources spent on this large area, its just some numbers.
